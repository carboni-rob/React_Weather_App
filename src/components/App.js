import React, { Component } from 'react';
import {
  Form,
  FormControl,
  FormGroup,
  Button,
  Grid,
  Row,
  Col
} from 'react-bootstrap';
import GoogleMapReact from 'google-map-react';
import ReactHighcharts from 'react-highcharts';
import { AsyncTypeahead } from 'react-bootstrap-typeahead';
import { bake_cookie, read_cookie } from 'sfcookies';

import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import { Card } from 'material-ui/Card';

import marker from '../img/marker2.png';

import styles from './styles';

const cookie_key = 'LOCATION';

const Marker = () => <img src={marker} alt="" style={styles.MARKER_STYLE} />;

class App extends Component {
  constructor() {
    super();
    this.state = {
      isQueryLoading: false,
      isDataLoading: false,
      dataLoaded: false,
      validationState: null,
      locationsArray: [],
      location: '',
      locationID: '',
      currentTemp: 'no data',
      currentWeather: 'no data',
      sundown: 'no data',
      center: { lat: 51.447501, lng: -2.60868 },
      zoom: 7,
      categories: [],
      maxTemps: [],
      avgTemps: [],
      minTemps: []
    };
  }

  componentDidMount() {
    var location = read_cookie(cookie_key);
    if (location.length !== 0) {
      this.setState({ location }, () => {
        this.onSubmit();
      });
    }
  }

  handleErrors(response) {
    if (!response.ok) {
      throw Error(response.statusText);
    }
    return response;
  }

  onSubmit() {
    this.setState({ isDataLoading: true });
    let categories = [];
    let maxTemps = [];
    let avgTemps = [];
    let minTemps = [];
    const { location } = this.state;
    const handleErrors = this.handleErrors;
    fetch(`http://interview.toumetisanalytics.com/location/${location}`)
      .then(handleErrors)
      .then(results => {
        return results.json();
      })
      .then(locationData => {
        let locationID = locationData[0].woeid;
        fetch(`http://interview.toumetisanalytics.com/weather/${locationID}`)
          .then(results => {
            return results.json();
          })
          .then(weatherData => {
            let currentTime = new Date(weatherData.time);
            let sunsetTime = new Date(weatherData.sun_set);
            if (sunsetTime > currentTime) {
              let msToSundow = new Date(sunsetTime - currentTime);
              let hoursToSundown = msToSundow.getHours();
              let minsToSundown = msToSundow.getMinutes();
              this.setState({
                sundown: `in ${hoursToSundown} Hours ${minsToSundown} Minutes`
              });
            } else {
              let msToSundow = new Date(currentTime - sunsetTime);
              let hoursToSundown = msToSundow.getHours();
              let minsToSundown = msToSundow.getMinutes();
              this.setState({
                sundown: `${hoursToSundown} Hours ${minsToSundown} Minutes ago`
              });
            }
            let lat_long = weatherData.latt_long.split(',');
            this.setState({
              center: {
                lat: Number(lat_long[0]),
                lng: Number(lat_long[1])
              }
            });
            this.setState({
              currentTemp: weatherData.consolidated_weather[0].the_temp.toFixed(
                2
              ),
              currentWeather:
                weatherData.consolidated_weather[0].weather_state_name,
              validationState: 'success'
            });
            weatherData.consolidated_weather.map(data => {
              var date = new Date(data.applicable_date);
              var day = date.getDate();
              var month = date.getMonth();
              var monthName;
              switch (month) {
                case 0:
                  monthName = 'Jan';
                  break;
                case 1:
                  monthName = 'Feb';
                  break;
                case 2:
                  monthName = 'Mar';
                  break;
                case 3:
                  monthName = 'Apr';
                  break;
                case 4:
                  monthName = 'May';
                  break;
                case 5:
                  monthName = 'Jun';
                  break;
                case 6:
                  monthName = 'Jul';
                  break;
                case 7:
                  monthName = 'Aug';
                  break;
                case 8:
                  monthName = 'Sep';
                  break;
                case 9:
                  monthName = 'Oct';
                  break;
                case 10:
                  monthName = 'Nov';
                  break;
                case 11:
                  monthName = 'Dec';
                  break;
                default:
                  monthName = 'Jan';
              }
              var dayMonth = `${day}. ${monthName}`;
              categories.push(dayMonth);
              var maxTemp = +data.max_temp.toFixed(2);
              var minTemp = +data.min_temp.toFixed(2);
              var avgTemp = +((maxTemp + minTemp) / 2).toFixed(2);
              maxTemps.push(maxTemp);
              avgTemps.push(avgTemp);
              minTemps.push(minTemp);
              return null;
            });
            this.setState({
              dataLoaded: true,
              categories,
              maxTemps,
              avgTemps,
              minTemps,
              isDataLoading: false
            });
            bake_cookie(cookie_key, this.state.location);
          });
      })
      .catch(() =>
        this.setState({ validationState: 'error', isDataLoading: false })
      );
  }

  errorMessage() {
    if (this.state.validationState === 'error') {
      return <h4 className="errMsg">Location not found</h4>;
    }
  }

  renderButton() {
    return (
      <Button
        disabled={this.state.isDataLoading}
        bsStyle="primary"
        onClick={() => {
          this.typeahead.getInstance().blur();
          this.onSubmit();
        }}>
        {this.state.isDataLoading ? 'Loading...' : 'Submit'}
      </Button>
    );
  }

  renderMap() {
    if (this.state.dataLoaded === true) {
      return (
        <Card className="mapCard" style={styles.cardStyle}>
          <h3>Last Location:</h3>
          <div
            style={{
              height: 350,
              width: '100%',
              display: 'flex',
              flexFlow: 'row nowrap',
              justifyContent: 'center',
              padding: 0
            }}>
            <GoogleMapReact
              bootstrapURLKeys={{
                key: 'AIzaSyCNNevvdpZC0MQ2fWOBgu7iqVkgNrN-IcU'
              }}
              center={this.state.center}
              defaultZoom={this.state.zoom}
              options={{
                panControl: false,
                mapTypeControl: false,
                scrollwheel: false
              }}>
              <Marker lat={this.state.center.lat} lng={this.state.center.lng} />
            </GoogleMapReact>
          </div>
        </Card>
      );
    }
  }

  renderChart() {
    var chartConfig = {
      chart: {
        type: 'column'
      },
      title: { text: '' },
      xAxis: {
        categories: this.state.categories
      },
      yAxis: {
        title: {
          text: 'Temperature'
        }
      },
      series: [
        {
          name: 'Max temp',
          data: this.state.maxTemps
        },
        {
          name: 'Avg temp',
          data: this.state.avgTemps
        },
        {
          name: 'Min temp',
          data: this.state.minTemps
        }
      ]
    };
    if (this.state.dataLoaded === true) {
      return (
        <Card className="chartCard" style={styles.cardStyle}>
          <h3>Six Day Forecast</h3>
          <ReactHighcharts config={chartConfig} />
        </Card>
      );
    }
  }

  populateAutocomplete(query) {
    this.setState({ isLoading: true });
    fetch(
      `https://cors-anywhere.herokuapp.com/http://metaweather.com/api/location/search/?query=${query}`
    )
      .then(results => {
        return results.json();
      })
      .then(locationsData => {
        var locationsArray = [];
        locationsData.map(location => {
          locationsArray.push(location.title);
          return null;
        });
        this.setState({ locationsArray, isQueryLoading: false });
      })
      .catch(error => console.log(error));
  }

  render() {
    return (
      <MuiThemeProvider>
        <div>
          <Grid>
            <Card style={styles.cardStyle}>
              <h2 style={{ margin: 10 }}>Weather Dashboard</h2>
            </Card>
            <Card style={styles.cardStyle}>
              <Form>
                <FormGroup
                  controlId="formValidation"
                  validationState={this.state.validationState}>
                  <AsyncTypeahead
                    ref={typeahead => (this.typeahead = typeahead)}
                    minLength={2}
                    isLoading={this.state.isQueryLoading}
                    options={this.state.locationsArray}
                    placeholder="Enter location"
                    onInputChange={query => {
                      this.setState({
                        location: query,
                        validationState: null
                      });
                      this.populateAutocomplete(query);
                    }}
                    onSearch={e => {
                      return null;
                    }}
                    onKeyDown={event => {
                      if (event.key === 'Enter') {
                        this.typeahead.getInstance().blur();
                        this.onSubmit();
                      }
                    }}
                  />
                  <FormControl.Feedback />
                  {this.errorMessage()}
                </FormGroup>
                {this.renderButton()}
              </Form>
            </Card>

            <Row className="show-grid">
              <Col xs={12} md={4}>
                <Card style={styles.cardStyle}>
                  <h3>Current Temp</h3>
                  <h4>{this.state.currentTemp}</h4>
                </Card>
              </Col>

              <Col xs={12} md={4}>
                <Card style={styles.cardStyle}>
                  <h3>Current Weather</h3>
                  <h4>{this.state.currentWeather}</h4>
                </Card>
              </Col>
              <Col xs={12} md={4}>
                <Card style={styles.cardStyle}>
                  <h3>Sundown</h3>
                  <h4>{this.state.sundown}</h4>
                </Card>
              </Col>
            </Row>
            <Row className="show-grid">
              <Col xs={12} md={4} />
              <Col xs={12} md={12}>
                {this.renderMap()}
                <Col xs={12} md={4} />
              </Col>
            </Row>
            {this.renderChart()}
          </Grid>
        </div>
      </MuiThemeProvider>
    );
  }
}

export default App;
