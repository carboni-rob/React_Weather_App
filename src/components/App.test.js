import React from 'react';
import App from './App';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

describe('App component', () => {
  it('matches the snapshot', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
