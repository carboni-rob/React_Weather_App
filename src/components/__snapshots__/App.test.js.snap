// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component matches the snapshot 1`] = `
<div>
  <div
    className="container"
  >
    <div
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0,0,0,0)",
          "backgroundColor": "#ffffff",
          "borderRadius": 2,
          "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12)",
          "boxSizing": "border-box",
          "color": "rgba(0, 0, 0, 0.87)",
          "fontFamily": "Roboto, sans-serif",
          "marginTop": 30,
          "muiPrepared": true,
          "padding": 20,
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "zIndex": 1,
        }
      }
    >
      <div
        style={
          Object {
            "paddingBottom": 0,
          }
        }
      >
        <h2
          style={
            Object {
              "margin": 10,
            }
          }
        >
          Weather Dashboard
        </h2>
      </div>
    </div>
    <div
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0,0,0,0)",
          "backgroundColor": "#ffffff",
          "borderRadius": 2,
          "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12)",
          "boxSizing": "border-box",
          "color": "rgba(0, 0, 0, 0.87)",
          "fontFamily": "Roboto, sans-serif",
          "marginTop": 30,
          "muiPrepared": true,
          "padding": 20,
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "zIndex": 1,
        }
      }
    >
      <div
        style={
          Object {
            "paddingBottom": 0,
          }
        }
      >
        <form
          className=""
        >
          <div
            className="form-group has-feedback"
          >
            <div
              className="rbt open clearfix"
              style={
                Object {
                  "position": "relative",
                }
              }
              tabIndex={-1}
            >
              <div
                className="rbt-input form-control"
                disabled={false}
                onClick={[Function]}
                onFocus={[Function]}
                tabIndex={-1}
              >
                <div
                  className="rbt-input-wrapper"
                >
                  <div
                    style={
                      Object {
                        "display": "block",
                        "position": "relative",
                      }
                    }
                  >
                    <input
                      aria-autocomplete="list"
                      aria-expanded={false}
                      aria-haspopup="true"
                      autoComplete="off"
                      className="rbt-input-main"
                      disabled={false}
                      name={undefined}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="Enter location"
                      role="combobox"
                      style={
                        Object {
                          "backgroundColor": "transparent",
                          "border": 0,
                          "boxShadow": "none",
                          "cursor": "inherit",
                          "outline": "none",
                          "padding": 0,
                          "width": "100%",
                        }
                      }
                      type="text"
                      value=""
                    />
                    <div
                      className={undefined}
                      style={
                        Object {
                          "bottom": 0,
                          "display": "block",
                          "pointerEvents": "none",
                          "position": "absolute",
                          "top": 0,
                          "zIndex": 0,
                        }
                      }
                    >
                      <input
                        aria-hidden={true}
                        className="rbt-input-hint"
                        style={
                          Object {
                            "backgroundColor": "transparent",
                            "border": 0,
                            "boxShadow": "none",
                            "boxSizing": "content-box",
                            "color": "rgba(0, 0, 0, 0.35)",
                            "cursor": "inherit",
                            "outline": "none",
                            "padding": 0,
                            "width": "1px",
                          }
                        }
                        tabIndex={-1}
                        value=""
                      />
                      <div
                        style={
                          Object {
                            "height": 0,
                            "left": 0,
                            "overflow": "scroll",
                            "position": "absolute",
                            "top": 0,
                            "visibility": "hidden",
                            "whiteSpace": "pre",
                          }
                        }
                      >
                        
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-atomic={true}
                aria-live="polite"
                className="sr-only rbt-sr-status"
                role="status"
              >
                0 selections
              </div>
            </div>
          </div>
          <button
            className="btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
    <div
      className="show-grid row"
    >
      <div
        className="col-md-4 col-xs-12"
      >
        <div
          style={
            Object {
              "WebkitTapHighlightColor": "rgba(0,0,0,0)",
              "backgroundColor": "#ffffff",
              "borderRadius": 2,
              "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12)",
              "boxSizing": "border-box",
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "Roboto, sans-serif",
              "marginTop": 30,
              "muiPrepared": true,
              "padding": 20,
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "zIndex": 1,
            }
          }
        >
          <div
            style={
              Object {
                "paddingBottom": 0,
              }
            }
          >
            <h3>
              Current Temp
            </h3>
            <h4>
              no data
            </h4>
          </div>
        </div>
      </div>
      <div
        className="col-md-4 col-xs-12"
      >
        <div
          style={
            Object {
              "WebkitTapHighlightColor": "rgba(0,0,0,0)",
              "backgroundColor": "#ffffff",
              "borderRadius": 2,
              "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12)",
              "boxSizing": "border-box",
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "Roboto, sans-serif",
              "marginTop": 30,
              "muiPrepared": true,
              "padding": 20,
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "zIndex": 1,
            }
          }
        >
          <div
            style={
              Object {
                "paddingBottom": 0,
              }
            }
          >
            <h3>
              Current Weather
            </h3>
            <h4>
              no data
            </h4>
          </div>
        </div>
      </div>
      <div
        className="col-md-4 col-xs-12"
      >
        <div
          style={
            Object {
              "WebkitTapHighlightColor": "rgba(0,0,0,0)",
              "backgroundColor": "#ffffff",
              "borderRadius": 2,
              "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12)",
              "boxSizing": "border-box",
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "Roboto, sans-serif",
              "marginTop": 30,
              "muiPrepared": true,
              "padding": 20,
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "zIndex": 1,
            }
          }
        >
          <div
            style={
              Object {
                "paddingBottom": 0,
              }
            }
          >
            <h3>
              Sundown
            </h3>
            <h4>
              no data
            </h4>
          </div>
        </div>
      </div>
    </div>
    <div
      className="show-grid row"
    >
      <div
        className="col-md-4 col-xs-12"
      />
      <div
        className="col-md-12 col-xs-12"
      >
        <div
          className="col-md-4 col-xs-12"
        />
      </div>
    </div>
  </div>
</div>
`;
